javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll

```diff
@@ -63,15 +64,24 @@ module ExceptionXss {
     NotYetThrown() { this = "NotYetThrown" }
   }
 
-  // Consider using "if (err) {.. [do something with err] .. }" as an extra condition if there are too many FP's.
+  /**
+   * A callback that is the last argument to some call, and the callback has the form:
+   * function (err, value) {if (err) {...} .. }
+   */
   class Callback extends DataFlow::FunctionNode {
+    DataFlow::ParameterNode errorParameter;
+
     Callback() {
       exists(DataFlow::CallNode call | call.getLastArgument().getAFunctionValue() = this) and
       this.getNumParameter() = 2 and
-      this.getParameter(0).getName().regexpMatch("err.*") // Using "e" was considered. But that matches too many jQuery methods where "element" is shortened as "e".
+      errorParameter = this.getParameter(0) and
+      exists(Expr errorCheck |
+        isExplicitConditional(this.getFunction().getBodyStmt(0), errorCheck) and
```

Comment:
While I appreciate the reuse, I think this will flag too many cases - why would you for instance check the error parameter in a loop condition?

How about the simpler, common case:
```
errorParameter.flowsToExpr(this.getFunction().getBodyStmt(0).(IfStmt).getCondition())
```

I also still think we need a reachability check that once the error-branch is taken, then we do not have a fall-through to the non-error branch. I do not have a reference at hand right now, but we do these reachability checks here and there using `BasicBlock` reasoning.



Future work with expected low payoffs:
- support `e != undefined`, `!!e` and friends (maybe `DefensiveProgramming::UndefinedNullTruthinessGuard` and friends are of some use)
- look into the prevalence of simple variable initialisation that happen *before* the if-statement.